encoding/multi_reader_iterator.go

```diff
@@ -0,0 +1,251 @@
+	return it.err != nil
+}
+
+func (it *multiReaderIterator) isClosed() bool {
+	return it.closed
+}
+
+func (it *multiReaderIterator) hasMore() bool {
+	return it.iters.Len() > 0 || it.slicesIter != nil
+}
+
+func (it *multiReaderIterator) hasNext() bool {
+	return !it.hasError() && !it.isClosed() && it.hasMore()
+}
+
+func (it *multiReaderIterator) moveToNext() {
+	if it.iters.Len() > 0 {
+		it.moveIteratorsToNext()
+	}
+	if it.iters.Len() > 0 || it.hasError() {
+		// Still have valid iters or has error
+		return
+	}
+
+	// Move forward through slices of readers
+	if !it.slicesIter.Next() {
+		// No more readers, nil out so that hasMore reflects correctly
+		it.slicesIter.Close()
+		it.slicesIter = nil
+		return
+	}
+
+	// Add all readers to current iterators heap
+	currentLen := it.slicesIter.CurrentLen()
+	for i := 0; i < currentLen; i++ {
+		var (
+			iter   m3db.ReaderIterator
+			reader = it.slicesIter.CurrentAt(i)
+		)
+		if len(it.cachedIters) != 0 {
+			iter = it.cachedIters[len(it.cachedIters)-1]
+			iter.Reset(reader)
+			it.cachedIters = it.cachedIters[:len(it.cachedIters)-1]
+		} else {
+			iter = it.iteratorAlloc(reader)
+		}
+		if iter.Next() {
+			// Only insert it if it has values
+			heap.Push(&it.iters, iter)
+		} else {
+			err := iter.Err()
+			iter.Close()
+			if err != nil {
+				it.err = err
+			}
+		}
+	}
+
+	if it.iters.Len() == 0 && !it.hasError() {
+		// No iterators were added, move to next
+		it.moveToNext()
+	}
+}
+
+func (it *multiReaderIterator) moveIteratorsToNext() {
+	iter := heap.Pop(&it.iters).(m3db.ReaderIterator)
+	prev, _, _ := iter.Current()
+
+	if it.moveIteratorToValidNext(iter) {
+		heap.Push(&it.iters, iter)
+	} else {
+		it.cachedIters = append(it.cachedIters, iter)
+	}
+
+	if it.iters.Len() == 0 {
+		return
+	}
+
+	curr, _, _ := it.Current()
+	if curr.Timestamp.Equal(prev.Timestamp) {
+		// Dedupe
+		it.moveIteratorsToNext()
+	}
+}
+
+func (it *multiReaderIterator) moveIteratorToValidNext(iter m3db.Iterator) bool {
+	prev, _, _ := iter.Current()
+	prevT := prev.Timestamp
+	if iter.Next() {
+		curr, _, _ := iter.Current()
+		t := curr.Timestamp
+		if t.Before(prevT) {
+			// Out of order datapoint
+			it.err = errOutOfOrderIterator
+			iter.Close()
+			return false
+		}
+		return true
+	}
+
+	err := iter.Err()
+	iter.Close()
+	if err != nil {
+		it.err = err
+	}
+	return false
+}
+
+func (it *multiReaderIterator) Err() error {
+	return it.err
+}
+
+func (it *multiReaderIterator) Reset(readers []io.Reader) {
+	it.singleSlicesIter.readers = readers
+	it.singleSlicesIter.clean = true
+	it.singleSlicesIter.closed = false
+	it.ResetSliceOfSlices(&it.singleSlicesIter)
+}
+
+func (it *multiReaderIterator) ResetSliceOfSlices(slicesIter m3db.ReaderSliceOfSlicesIterator) {
+	it.iters = it.iters[:0]
+	it.slicesIter = slicesIter
+	it.err = nil
+	it.clean = true
+	it.closed = false
+	heap.Init(&it.iters)
+	// Try moveToNext to load values for calls to Current before Next
+	it.moveToNext()
```

Comment:
Yeah so there's actually a lot of code that relies on being able to call "Current()" before "Next()", I would advocate to keep this logic which is currently consistent across all our iterators.